diff --git a/man/waybar-custom.5.scd b/man/waybar-custom.5.scd
index 67e4c89c..2f34e041 100644
--- a/man/waybar-custom.5.scd
+++ b/man/waybar-custom.5.scd
@@ -123,11 +123,11 @@ This should look like this:
 
 The *class* parameter also accepts an array of strings.
 
-If nothing or an invalid option is specified, Waybar expects i3blocks style output. Values are *newline* separated.
+If nothing or an invalid option is specified, Waybar expects *Tab* delemited output.
 This should look like this:
 
 ```
-$text\\n$tooltip\\n$class*
+$text\\t$tooltip\\t$class\\n*
 ```
 
 *class* is a CSS class, to apply different styles in *style.css*
diff --git a/src/modules/custom.cpp b/src/modules/custom.cpp
index fa03c3be..997a0819 100644
--- a/src/modules/custom.cpp
+++ b/src/modules/custom.cpp
@@ -195,7 +195,7 @@ void waybar::modules::Custom::parseOutputRaw() {
   std::istringstream output(output_.out);
   std::string line;
   int i = 0;
-  while (getline(output, line)) {
+  while (getline(output, line, '\t')) {
     Glib::ustring validated_line = line;
     if (!validated_line.validate()) {
       validated_line = validated_line.make_valid();
